### ISSUE TITLE
SocketException (104): Connection reset by peer

### ISSUE THREAD

**Issue by Henery309** [microsoft/reverse-proxy#921]
### Describe the bug
Yarp proxy is in prod - serving 500-1k requests per minute. Looking at the logs we are constantly getting the error below.
Happens 4-5 times a day. It is random we can't reproduce it. Hoping you guys could help. We were looking to move proxy to serve higher traffic website but this is blocking.

Instance is hosted in dotnet 5 using aspnet:5.0-buster-slim image on linux.

### Further technical details

      at Microsoft.ReverseProxy.Service.Proxy.HttpProxy.ProxyAsync(HttpContext context, String destinationPrefix, HttpMessageInvoker httpClient, RequestProxyOptions requestOptions, HttpTransformer transformer)
      at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
      at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
      --- End of inner exception stack trace ---
      at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
      at System.Net.Http.Http2Connection.Http2Stream.ReadResponseHeadersAsync(CancellationToken cancellationToken)
      at System.Net.Http.Http2Connection.Http2Stream.TryEnsureHeaders()
      at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
      at System.Net.Http.Http2Connection.ThrowRequestAborted(Exception innerException)
      --- End of inner exception stack trace ---
      at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
      at System.Net.Security.SslStream.ReadAsyncInternal[TIOAdapter](TIOAdapter adapter, Memory`1 buffer)
      at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token)
      at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
      End of inner exception stack trace 
      --- System.Net.Sockets.SocketException (104): Connection reset by peer
      --- System.IO.IOException: Unable to read data from the transport connection: Connection reset by peer.
      --- System.IO.IOException: The request was aborted.

**Comment by Tratcher** [comment]
It's interesting that an HTTP/2 connection is getting abortively closed before receiving the response headers. Are there any matching logs on the destination server about the connection being terminated?

Ideally you'd be able to collect a network trace to clarify what happened to the TCP connection, but I can appreciate that being difficult for something that only reproduces under load.

Can you summarize the kind of requests you're proxying? Are you proxying any specific kinds of traffic like gRPC? The stack shows you're using HTTP/2 and TLS to communicate with the destination. If you have leeway to experiment you could try using HTTP/1.1 instead.
https://github.com/microsoft/reverse-proxy/blob/e54d6c31778b7da2910fff9836832a9cfc82494f/src/ReverseProxy/Service/Proxy/RequestProxyOptions.cs#L24
https://github.com/microsoft/reverse-proxy/blob/e54d6c31778b7da2910fff9836832a9cfc82494f/src/ReverseProxy/Abstractions/ClusterDiscovery/Contract/Cluster.cs#L47

**Comment by Henery309** [comment]
Hi @Tratcher I forgot that http/2 was enabled by default in Yarp. All my destination services are aspnet core 5 rest APIs. I will change yarp to http/1.1 and come back to you.

Following is a request flow for my application.
`Https GCP load balancer -> YARP proxy (linux vm) -> Gcp Cloud run (http/2 disabled). AspNet core 5 web API.
`

Also wondering why is http/2 enabled by default in yarp? My understanding is that http/2 is mostly used with Grpc.

**Comment by Tratcher** [comment]
> Also wondering why is http/2 enabled by default in yarp? My understanding is that http/2 is mostly used with Grpc.

Yes, we do want gRPC to work by default. HTTP/2 also provides some density benefits for the proxy (fewer connections, more efficient wireformat, etc).

I suggested turning it off only to see if it affects the repro. Your diagram says your destinations have HTTP/2 disabled, but your logs say otherwise.

**Comment by Henery309** [comment]
Hi @Tratcher today ran the proxy under http 1.1 and received the following error. 

```
at Microsoft.ReverseProxy.Service.Proxy.HttpProxy.ProxyAsync(HttpContext context, String destinationPrefix, HttpMessageInvoker httpClient, RequestProxyOptions requestOptions, HttpTransformer transformer)
at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
--- End of inner exception stack trace ---
at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
at System.Net.Security.SslStream.ReadAsyncInternal[TIOAdapter](TIOAdapter adapter, Memory`1 buffer)
at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token)
at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken
--- End of inner exception stack trace ---
---> System.Net.Sockets.SocketException (104): Connection reset by peer
---> System.IO.IOException: Unable to read data from the transport connection: Connection reset by peer.
```

**Comment by Tratcher** [comment]
Hmm, that's the same underlying socket error. Is it happening with a similar frequency as when HTTP/2 was enabled?

According to both of those stacks the abort is initiated by the destination server. You're going to have to dig into the logs there to get more information.

**Comment by karelz** [comment]
Triage: Server side logs are needed for root cause.

**Comment by Henery309** [comment]
Unfortnely its not possible to get network trace from a GCP cloud run (serverless hosting). I am going to try deploy to Azure container service. I will notify if same issue occurs on azure.

**Comment by karelz** [comment]
Triage: No update, closing. Feel free to reopen if there is something actionable for us to look at. Thanks!
